added tests  8a & 8b to do 1D/2D mesh writing and reading

guj opened the pull request.

Commit:
added tests  8a & 8b to do 1D/2D mesh writing and reading

Commit:
tabs fix

Commit:
space fix

Commit:
added comments and defaults

Commit:
style fix

Commit:
more style fix

Commit:
added position/offset for validation requirement

Commit:
fixed a typo plus format change

Commit:
fixed a bug & fixed typos

Commit:
tab

Commit:
eol fix

ax3l commented on examples/8a_benchmark_write1D2D_parallel.cpp:
```diff
@@ -0,0 +1,651 @@
+public:
+  TestInput() =  default;
+
+  unsigned int GetSeg() const
```
Please add a doxygen string to this function.

ax3l commented on examples/8a_benchmark_write1D2D_parallel.cpp:
```diff
@@ -0,0 +1,651 @@
+/**     Class TestInput
+ *
+ *
+ * @param mpi_size      MPI size
+ * @param mpi_rank      MPI rank
+ * @param bulk          num of elements
+ * @param numSeg        num of subdivition for the elements
+ *                      note that with h5collect mode, numSeg must be 1
+ * @param numSteps      num of iterations
```
I might overlook this, but do these parameters belong here? `@param` is usually a function parameter. Let me document below how to describe member variables.

ax3l commented on examples/8a_benchmark_write1D2D_parallel.cpp:
```diff
@@ -0,0 +1,651 @@
+  }
+
+
+  int m_MPISize = 1;
+  int m_MPIRank = 0;
+  unsigned long m_Bulk = 1000ul;
+  unsigned int m_Seg = 1;
+  int m_Steps = 1;
+  int m_TestNum = 0;
+  std::string m_Backend = ".bp";
+  bool m_Unbalance = false;
+
+  int m_Ratio = 1;
+
+  Extent m_GlobalMesh;
+  std::vector<std::pair<unsigned long, unsigned long>> m_InRankDistribution;
```
```suggestion
  int m_MPISize = 1; //!< MPI size
  int m_MPIRank = 0; //!< MPI rank
  unsigned long m_Bulk = 1000ul; //!< num of elements
  /** number of subdivisions for the elements
   *
   * note that with h5collect mode, m_Seg must be 1
   */
  unsigned int m_Seg = 1;
  int m_Steps = 1;   //!< num of iterations
  int m_TestNum = 0; //!< ... description here ...
  std::string m_Backend = ".bp"; //!< I/O backend by file ending
  bool m_Unbalance = false;      //!< ... description here ...

  int m_Ratio = 1; //!< ... description here ...

  Extent m_GlobalMesh; //!< ... description here ...
  /** ... description here ...
   */
  std::vector<std::pair<unsigned long, unsigned long>> m_InRankDistribution;
```

ax3l commented on examples/8a_benchmark_write1D2D_parallel.cpp:
```diff
@@ -0,0 +1,651 @@
+        input.m_Steps = atoi( argv[4] );
+
+
+    auto backends = getBackends();
+    for (auto which: backends)
+      {
+    input.m_Backend = which;
+    input.run(1);
+    input.run(2);
+      }
```
```suggestion
    auto const backends = getBackends();
    for( auto const & which: backends )
    {
        input.m_Backend = which;
        input.run(1);
        input.run(2);
    }
```

ax3l commented on examples/8b_benchmark_read1D2D_parallel.cpp:
```diff
@@ -0,0 +1,519 @@
+
+class MemoryProfiler
+{
+    /**
+     *
+     * Simple Memory profiler for linux
+     *
+     * @param tag      item name to measure
+     * @param rank     MPI rank
+     */
+public:
```
```suggestion
public:
    /** Simple Memory profiler for linux
     *
     * @param[in] rank     MPI rank
     * @param[in] tag      item name to measure
     */
```

ax3l commented on examples/8b_benchmark_read1D2D_parallel.cpp:
```diff
@@ -0,0 +1,519 @@
+#endif
+  }
+
+    /**
+     *
+     * Read from /proc/self/status and display the Virtual Memory info at rank 0 on console
+     *
+     * @param tag      item name to measure
+     * @param rank     MPI rank
+     */
+
```
```suggestion
    /** Read from /proc/self/status and display the Virtual Memory info at rank 0 on console
     *
     * @param tag      item name to measure
     */
```

ax3l commented on CMakeLists.txt:
```diff
@@ -627,6 +627,9 @@ set(openPMD_EXAMPLE_NAMES
     8_benchmark_parallel
+    8a_benchmark_write1D2D_parallel
+    8b_benchmark_read1D2D_parallel
+
```
```suggestion
```

guj replied:
ok

ax3l commented on examples/8a_benchmark_write1D2D_parallel.cpp:
```diff
@@ -0,0 +1,651 @@
+
+class MemoryProfiler
+{
+    /**
+     *
+     * Simple Memory profiler for linux
+     *
+     * @param tag      item name to measure
+     * @param rank     MPI rank
+     */
+public:
```
```suggestion
public:
    /** Simple Memory profiler for linux
     *
     * @param[in] rank     MPI rank
     * @param[in] tag      item name to measure
     */
```

ax3l replied:
Maybe a `std::string` for the tag can be more C++-ish?

ax3l commented on examples/8a_benchmark_write1D2D_parallel.cpp:
```diff
@@ -0,0 +1,651 @@
+#endif
+  }
+
+    /**
+     *
+     * Read from /proc/self/status and display the Virtual Memory info at rank 0 on console
+     *
+     * @param tag      item name to measure
+     * @param rank     MPI rank
+     */
+
```
```suggestion
    /** Read from /proc/self/status and display the Virtual Memory info at rank 0 on console
     *
     * @param[in] tag      item name to measure
     */
```

ax3l commented on examples/8a_benchmark_write1D2D_parallel.cpp:
```diff
@@ -0,0 +1,651 @@
+
+static std::chrono::time_point<std::chrono::system_clock>  m_ProgStart = std::chrono::system_clock::now();
+
+class MemoryProfiler
```
Should  this class go into `include/openPMD/benchmark/`? That would reduce some code duplication in both examples.

ax3l commented on examples/8b_benchmark_read1D2D_parallel.cpp:
```diff
@@ -0,0 +1,519 @@
+    if (argc >= 3)
+      input.m_Pattern = atoi(argv[2]);
+
+    auto backends = getBackends();
+    for (auto which: backends)
+      {
+    input.m_Backend = which;
+    input.run(prefix);
+      }
```
```suggestion
    auto const backends = getBackends();
    for( auto const & which: backends )
    {
        input.m_Backend = which;
        input.run(prefix);
    }
```

guj committed:
Update examples/8a_benchmark_write1D2D_parallel.cpp

style change by Axel

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

guj committed:
Update examples/8a_benchmark_write1D2D_parallel.cpp

style change by Axel..

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

guj committed:
Update examples/8a_benchmark_write1D2D_parallel.cpp

style change by Axel

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

guj committed:
Update examples/8b_benchmark_read1D2D_parallel.cpp

style change by Axel ...

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

Commit:
more comments
removed const char*

Commit:
replaced char* with string

Commit:
eol

ax3l committed:
Cleanup Benchmarks

Deduplicate auxiliary code, fix formatting, fix doxygen, try to
document undocumented parameters.
